### ISSUE TITLE
Failed to start workspace with SSL: Server 'theia' in container 'theia-idei8u' not available

### ISSUE THREAD

**Issue by josh234098324** [eclipse-che/che#18062]
Runtime
v1.18.6

Che version:
Che Version: 7.18.1

Steps to reproduce
Deployed Eclipse Che using helm with the attached values files in my k8s environment:

Installation method:
helm template che -f che_master_values.yaml -n che | k apply (see attached for contents)
helm template nginx -f ingress_0.35_values.yaml -n nginx | k apply (see attached for contents)

Pulled up the Che UI and went to "Get started," clicked temporary storage and then clicked C/C++. This showed the below error when it started.

Describe the bug:
[cpp-pluginsl97] -> Deploying backend plugin "cdt-gdb-vscode@0.0.91" from "/tmp/vscode-unpacked/che-incubator.cpptools.latest.fbbpmsgadc.cdt-gdb-vscode-0.0.91-2cbbb8.vsix/extension/out/extension"
[cpp-pluginsl97] -> Deploying backend plugin "vscode-clangd@0.1.5" from "/tmp/vscode-unpacked/che-incubator.cpptools.latest.lrstmlodji.vscode-clangd-0.1.5-562d00.vsix/extension/out/src/extension"
[cpp-pluginsl97] -> Deploying backend plugin "cdt-vscode@0.0.7" from "/tmp/vscode-unpacked/che-incubator.cpptools.latest.osgkhqisdd.cdt-vscode-0.0.7-75cf95.vsix/extension/dist/extension.js"
[cpp-pluginsl97] -> Deploy plugins list took: 1145.1 ms
Error: Failed to run the workspace: "Server 'theia' in container 'theia-idedrn' not available."

##Che Pod Log:

2020-10-06 17:02:19,917[nio-8080-exec-4]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 480]   - Starting workspace 'che/cpp-zqj6t' with id 'workspace4pkg6053xvmtluty' by user 'che'
2020-10-06 17:02:20,125[aceSharedPool-0]  [WARN ] [i.f.k.c.i.VersionUsageUtils 60]      - The client is using resource type 'ingresses' with unstable version 'v1beta1'
2020-10-06 17:06:26,125[aceSharedPool-0]  [WARN ] [.i.k.KubernetesInternalRuntime 247]  - Failed to start Kubernetes runtime of workspace workspace4pkg6053xvmtluty. Cause: Server 'theia' in container 'theia-idei8u' not available.
2020-10-06 17:06:27,714[aceSharedPool-0]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 984]   - Workspace 'che:cpp-zqj6t' with id 'workspace4pkg6053xvmtluty' start failed

Che Workspace Ingresses:
k get ing -n che-che        
NAME              CLASS    HOSTS                  ADDRESS     PORTS     AGE
ingress6qtyp3br   <none>   che.\<domain\>           \<ip\>       80, 443   93s
ingressfes1rmct   <none>   che.\<domain\>           \<ip\>       80, 443   93s
ingressg3nzushp   <none>   che.\<domain\>           \<ip\>       80, 443   93s
ingresskqnw2twj   <none>   che.\<domain\>           \<ip\>       80, 443   93s
ingressuevyqsva   <none>   che.\<domain\>           \<ip\>       80, 443   93s
ingressvptzt1ef   <none>   che.\<domain\>           \<ip\>       80, 443   93s
ingresszpkx9apx   <none>   che.\<domain\>           \<ip\>       80, 443   93s

Nginx pod logs:
"GET /api/websocket HTTP/1.1" 101 
Service "che-che/server6g7t9jo4-theia-ide6wz" does not have any active Endpoint.
Service "che-che/server6g7t9jo4-theia-ide6wz" does not have any active Endpoint.
Service "che-che/server6g7t9jo4-theia-ide6wz" does not have any active Endpoint.
Service "che-che/serverzhiptekz-jwtproxy" does not have any active Endpoint.
Service "che-che/server6g7t9jo4-theia-ide6wz" does not have any active Endpoint.
Service "che-che/serverzhiptekz-jwtproxy" does not have any active Endpoint.
Service "che-che/serverzhiptekz-jwtproxy" does not have any active Endpoint.
"GET /api/workspace/workspace4pkg6053xvmtluty HTTP/1.1" 200 
"GET /api/preferences?filter=theia-user-preferences HTTP/1.1" 200
Service "che-che/serverzhiptekz-jwtproxy" does not have any active Endpoint.
Service "che-che/serverzhiptekz-jwtproxy" does not have any active Endpoint.
Service "che-che/serverzhiptekz-jwtproxy" does not have any active Endpoint.

Logs for the Theia container show what seems like a successful startup and then about 4-5 minutes later receives a SIGTERM. It is exactly 4.5 minutes for the attempting to connect to the workspace portion.  

Browsers tested with: Chrome, Firefox, EDGE (Win 10)

What worked
The Che dashboard works fine along with the devfile and plugin registries. Kubernetes is able to pull up the necessary pods and talk over SSL to the other services. I'm able to go to "https:\/\/che.\<domain\>\/\<pathforIDE\>" and it pulls up the IDE fairly quickly and shows the connection as secured. Also, using the same load balancer I'm able to pull up the workspaces properly when I disable SSL in Che and Nginx. 

Expected behavior
What I think should happen is the dashboard should be able to pull up the workspace inside the same frame the logs show up when starting the container.

I've tried all different versions of every plugin and editor. I'm on the Che channel in Mattermost under the name "josh" as well.
[che_master_values.txt](https://github.com/eclipse/che/files/5336426/che_master_values.txt)
[ingress_0.35_values.txt](https://github.com/eclipse/che/files/5336428/ingress_0.35_values.txt)

**Comment by tolusha** [comment]
@josh234098324 
Basically it should work. What does print? 

`kubectl describe ingress -n che-che`
`kubectl get secrets -n che-che`

**Comment by josh234098324** [comment]
See attached. 
[ing_che-che.txt](https://github.com/eclipse/che/files/5341570/ing_che-che.txt)
[che_che_secrets.txt](https://github.com/eclipse/che/files/5341571/che_che_secrets.txt)

**Comment by tolusha** [comment]
To make sure that there is no issue with certificates could you try
`curl -v --cacert <cacert_file> <> https://<domain>/serverhcychrl4-jwtproxy/server-4402/` (`serverhcychrl4-jwtproxy` will be different for a new workspace)
where `cacert_file` with a content of the `self-signed-certificate`  secret.

**Comment by josh234098324** [comment]
The output for items 1 and 2 are in the attached "che_curl_and_container_output.txt" and an attempt to connect in a browser is in the image below. It shows the lock for the site being secure but since I removed the domain it doesn't show in the JPG. For item three, the certificate is setup for "*.<domain>" which can be seen in the curl output. 

With the ingressDomain set to just <domain> and the cheHost set to just che, che.<domain> or commented out it deploys all the services with just the domain set as the host so none of the services are accessible. 

When doing the curl if the / at the end is removed from the URL it redirects to /dashboard and it does the same in the browser. Could it be possible that somehow the / is being removed and it just keeps being redirected to /dashboard instead of the jwtproxy and, if so, what could cause something like that to happen? 

![theia_site](https://user-images.githubusercontent.com/70341046/95467220-d15da980-0942-11eb-8903-8d4082700ede.JPG)
[che_curl_and_container_output.txt](https://github.com/eclipse/che/files/5348257/che_curl_and_container_output.txt)

**Comment by tolusha** [comment]
So, Theia is accessible from the browser and by curl but che-server can't access it ^(
Could you try comment out `cheHost` and try to deploy one more time with ingressDomain = <domain> ?

**Comment by tolusha** [comment]
> Could it be possible that somehow the / is being removed and it just keeps being redirected to /dashboard instead of the jwtproxy and, if so, what could cause something like that to happen?

No, I don't think so.

**Comment by josh234098324** [comment]
> So, Theia is accessible from the browser and by curl but che-server can't access it ^(
> Could you try comment out `cheHost` and try to deploy one more time with ingressDomain = ?

When you say "?" do you mean \<domain\> only or che.\<domain\>?

**Comment by tolusha** [comment]
I've tried with and without `che.` prefix using the helm charts  [1].
In both cases I deployed Eclipse Che successfully (on minikube). 
[1] https://github.com/eclipse/che/tree/master/deploy/kubernetes/helm/che

Some thoughts, are certificates from `self-signed-certificate`  and `che-tls` (ca.crt key) secrets are two different certificates?

**Comment by josh234098324** [comment]
Deploying with no host and no "che" at the beginning of the ingress just made it to where I couldn't pull up the Load Balancer since it was just the domain on everything which isn't a page that's presented. I went in and added "che." to the ingresses and configuration CM for Che to use "che.\<domain\>" instead and I was able to get to the site again but still produced the same results. 

Just for clarification my DNS entry for the load balancer is the "che.\<domain\>" so the URL is "https://che.\<domain\>\dashboard" and yes the curl command was ran from the Theia container though I tried from other containers that trusted the same certificates and it worked as well. 

The self-signed-certificate contains the PEM of the approving certificates. The che-tls contains the key and the crt that the Ingress uses. I originally didn't have a ca.crt entry in the secret but tried just now and that didn't help either. 

I was able to keep the workspace alive for the past hour by opening random things inside the IDE but the dashboard just kept spinning and waiting.

**Comment by tolusha** [comment]
That's something I don't understand.

`self-signed-certificate` should contain only CA certificates (one or chain).
`che-tls` should contain a server private key and a server certificate.
so `ca.crt` (from `self-signed-certificate`) != `ca.crt` (from che-tls)

Could you confirm that everything is correct on your side?

**Comment by josh234098324** [comment]
ca.crt (from self-signed-certificate) = PEM of approving certificate chain

cht-tls = Server certificate and Server private key (I only attempted adding the ca.crt in hopes it might help)

**Comment by tolusha** [comment]
@josh234098324 
- could you try to reach theia endpoint from the che-server container?

> I only attempted adding the ca.crt in hopes it might help

here I would recommend to create all secrets and only then redeploy Eclipse Che from the begining.

**Comment by josh234098324** [comment]
The che-server container doesn't have curl so I had to use a sidecar in order to do it but it was able to connect to the webpage with the certs in the secret. And I deployed Che again with the namespace and secrets already created and it didn't make a difference. 

I noticed the error below when I pulled up another instance that I didn't notice before because the welcome page covered it up but could this error be related to my issues and, if so, what would I look at to fix it? 

"Cannot GET /servern6234dh2-jwtproxy/server-4402/webview/index.html"
